semantics/src/main/kotlin/com/strumenta/kolasu/semantics/indexing/Index.kt

```diff
@@ -0,0 +1,70 @@
+package com.strumenta.kolasu.semantics.indexing
+
+/**
+ * Abstract storage component for external symbols.
+ **/
+interface Index {
```

Comment:
Should this be called SymbolIndex?

Reply:
Perhaps the documentation comment is confusing, the rationale behind the index interface is: to have a component responsible for storing external information/symbols/data, e.g. symbols, types, built-in things not stored in any AST, etc. Anything that should be visible from an AST but is not included in that AST. 

Given the rationale, I would not refer to Symbols only in this case. Using the index to store external symbols is one possible usage of this data structure. 

If this makes development easier somewhere else, we could perhaps define a SymbolIndex interface extending this one. But I am not sure about the advantage of this, any specific use case in mind?

Reply:
Maybe then it's not really an _Index_ but an _Environment_ or similar concept?
_Index_ to me suggests something that's optimized for certain search queries, e.g. partitioning the values into some sort of ordered tree.

Reply:
> Maybe then it's not really an _Index_ but an _Environment_ or similar concept? _Index_ to me suggests something that's optimized for certain search queries, e.g. partitioning the values into some sort of ordered tree.

It is meant to be an interface where to store views of data, which details depend on the specific objective for which we are storing it. For example, we could have entries for node positions consisting of an identifier and a position value, or else. 

Aren't the specifics of data partitioning and query execution something that would belong to a specific implementation of the Index?

I would be fine with using another name like _Environment_, but then it might become unclear that we can also store partial views of things using custom index writer configurations. The point is that whether it is an environment, a cache or else depends on the implementation you choose or define and the writer configuration.

Reply:
To me seems an over-generalized abstraction. My suggestion is to create generic abstractions when you have 3 different concrete cases. Without concrete examples the interface seems too "abstract" to me. I would not know when to use it and when not to use it. To me it seems a subset of the `Collection` interface.

Reply:
I understand this index is a portal to the external world. So in my mind I'd call it `ExternalStorage` or something like that. And it makes sense that in the library there is only a "trivial" implementation of it, because implementations using LSIF or the LW Model Repo etc. do not belong in Kolasu Core. This is I think what other libraries/frameworks call an "SPI" (e.g. Hibernate), i.e. an extension point that you can inject into the library. Kind of the opposite direction of an API, that is meant to be called by consumers of the library, while the SPI is meant to be called by the library with a well-defined protocol. 
It makes sense to me now, but I was initially confused by the name "Index" that maybe comes from LSIF.
That said, I too think that index reader and writer are maybe a bit overkill, but I don't think that should stop this PR from being merged; we can always simplify the design later, if the API proves to be confusing, verbose, etc.

Reply:
After discussion we decided to rename the package persistence and this class `Repository`